contracts/Governance.sol

```diff
@@ -211,23 +213,152 @@ contract Governance is Context, Ownable, ERC165, EIP712 {
+    {
         _votingDelay = votingDelay_;
     }
 
-    function setVotingPeriod(uint256 votingPeriod_) external virtual onlyOwner {
+    function setVotingPeriod(uint256 votingPeriod_)
+        external
+        virtual
+        onlyGovernance
+    {
         _votingPeriod = votingPeriod_;
     }
 
+    function registerProtocol(
+        address protocol,
+        bytes32 role,
+        uint256 quorum
+    ) external virtual onlyGovernance {
+        protocolsRoles[protocol] = role;
+        protocolsQuorums[protocol] = quorum;
+    }
+
+    function unregisterProtocol(address protocol)
+        external
+        virtual
+        onlyGovernance
+    {
+        delete protocolsRoles[protocol];
+        protocolsQuorums[protocol] = 0;
+    }
+
+    function registerRole(bytes32 role) external virtual onlyGovernance {
+        rolesList.push(role);
+        _roles[role] = rolesList.length;
+    }
+
+    function unregisterRole(bytes32 role) external virtual onlyGovernance {
+        rolesList[_roles[role]] = rolesList[rolesList.length - 1];
+        _roles[rolesList[rolesList.length - 1]] = _roles[role];
+        delete rolesList[rolesList.length - 1];
+        _roles[role] = 0;
+    }
+
+    function registerAction(
+        address target,
+        string calldata signature,
+        bytes32 role,
+        uint256 quorum
+    ) external virtual roleExists(role) onlyGovernance {
+        actionsRoles[target][signature] = role;
+        actionsQuorums[target][signature] = quorum;
+    }
+
+    function unregisterAction(
+        address target,
+        string calldata signature,
+        bytes32 role
+    ) external virtual roleExists(role) onlyGovernance {
+        delete actionsRoles[target][signature];
+        actionsQuorums[target][signature] = 0;
+    }
+
+    function addRoleMember(
+        bytes32 role,
+        address member,
+        address proposer
+    ) external virtual roleExists(role) onlyGovernance {
+        require(
+            votersRoles[role][proposer],
+            'Governance::addRoleMember: proposer must have the same role'
+        );
+        votersRoles[role][member] = true;
+    }
+
+    function removeRoleMember(
+        bytes32 role,
+        address member,
+        address proposer
+    ) external virtual roleExists(role) onlyGovernance {
+        require(
+            votersRoles[role][proposer],
+            'Governance::removeRoleMember: proposer must have the same role'
+        );
+        votersRoles[role][member] = false;
+    }
+
+    function setProposalThreshold(
+        bytes32 role,
+        uint256 threshold,
+        address proposer
+    ) external virtual roleExists(role) onlyGovernance {
+        require(
+            votersRoles[role][proposer],
+            'Governance::setProposalThreshold: proposer must have the same role'
+        );
+        proposalThresholds[role] = threshold;
+    }
+
+    function setVoterRolesAdmin(address voter, bytes32[] calldata voterRoles)
```

Comment:
question: this function name may be misleading. Is the purpose of the function to assign the admin role to the given voter?

Reply:
No, it allows contract owner to set roles for voter initially as we need some initial set of roles for voters.

Reply:
suggestion: as this sounds like an init operation, move to either an `init` or the constructor.

If you want to keep the function, I'd suggest renaming to something like `initVoterRoles` as the function init's the voter roles.